File: SPCode/CS/UI/RadioButton/SPRadioButtonList.cs
using System.Text.RegularExpressions;
using System.Diagnostics;
using System;
using System.Drawing.Drawing2D;
using System.Windows.Forms;
using System.Collections;
using System.Drawing;
using Microsoft.VisualBasic;
using System.Data.SqlClient;
using System.Data;
using System.Collections.Generic;
using WeifenLuo.WinFormsUI;
using Microsoft.Win32;
using WeifenLuo;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing.Design;
using System.Windows.Forms.Design;
using System.Windows.Forms.ComponentModel;


namespace SoftLogik.Win
{
	namespace UI
	{
		[Designer(typeof(SPRadioButtonListDesigner))][ToolboxBitmap(typeof(SPRadioButtonListDesigner))][DefaultEvent("SelectedIndexChanged")]public partial class SPRadioButtonList
		{
			
			
			public enum RadioLayoutStyles
			{
				@Horizontal,
				@Vertical,
				@Table
			}
			
			public delegate void SelectedIndexChangedEventHandler(object sender, System.EventArgs e);
			private SelectedIndexChangedEventHandler SelectedIndexChangedEvent;
			
			public event SelectedIndexChangedEventHandler SelectedIndexChanged
			{
				add
				{
					SelectedIndexChangedEvent = (SelectedIndexChangedEventHandler) System.Delegate.Combine(SelectedIndexChangedEvent, value);
				}
				remove
				{
					SelectedIndexChangedEvent = (SelectedIndexChangedEventHandler) System.Delegate.Remove(SelectedIndexChangedEvent, value);
				}
			}
			
			
			private SPRadioButtonItemCollection _RadioList = new SPRadioButtonItemCollection();
			private RadioLayoutStyles _radioLayoutStyle;
			private int _SelectedIndex = - 1;
			private object _SelectedValue = null;
			
			#region Overrides
			protected override void OnLoad(System.EventArgs e)
			{
				base.OnLoad(e);
				
				try
				{
					this.RadioGroupBox.Text = this.Text;
					BuildRadioTable();
				}
				catch (Exception)
				{
				}
			}
			
			protected void OnCheckedChanged(System.Object sender, System.EventArgs e)
			{
				
			}
			
			#endregion
			
			#region Properties
			[Category("Behavior")][Description("Gets or Sets the Radio Buttons Layout Style")]public RadioLayoutStyles LayoutStyle
			{
				get
				{
					return _radioLayoutStyle;
				}
				set
				{
					_radioLayoutStyle = value;
				}
			}
			[Category("Behavior"), Browsable(true), EditorAttribute(typeof(SPRadioButtonListItemsEditor), typeof(System.Drawing.Design.UITypeEditor))]public SPRadioButtonItemCollection Items
			{
				get
				{
					return this._RadioList;
				}
				set
				{
					this._RadioList = value;
					BuildRadioTable();
				}
			}
			#endregion
			
			#region Methods
			public void Add(string Name, string Text, bool Selected)
			{
				this._RadioList.Add(new SPRadioButtonItem(Name, Text, Selected));
			}
			
			public void Remove(string Name)
			{
				this._RadioList.Remove(this._RadioList[Name]);
			}
			#endregion
			
			#region Data Binding
			private CurrencyManager m_currencyManager = null;
			private string m_ValueMember;
			private string m_DisplayMember;
			private object m_oDataSource;
			
			[Category("Data")]public object DataSource
			{
				get
				{
					return m_oDataSource;
				}
				set
				{
					if (value == null)
					{
						this.m_currencyManager = null;
						this.Controls.Clear();
					}
					else
					{
						if (!(value is IList|| m_oDataSource is IListSource))
						{
							throw (new System.Exception("Invalid DataSource"));
						}
						else
						{
							if (value is IListSource)
							{
								IListSource myListSource = (IListSource) value;
								if (myListSource.ContainsListCollection == true)
								{
									throw (new System.Exception("Invalid DataSource"));
								}
							}
							this.m_oDataSource = value;
							this.m_currencyManager = (CurrencyManager) (this.BindingContext[value]);
							BuildRadioTable();
						}
					}
				}
			} // end of DataSource property
			
			[Category("Data")]public string ValueMember
			{
				get
				{
					return this.m_ValueMember;
				}
				set
				{
					this.m_ValueMember = value;
				}
			}
			[Category("Data")]public string DisplayMember
			{
				get
				{
					return this.m_DisplayMember;
				}
				set
				{
					this.m_DisplayMember = value;
				}
			}
			public object GetValue(int index)
			{
				IList innerList = this.m_currencyManager.List;
				if (innerList != null)
				{
					if ((this.ValueMember != "") && (index >= 0 && 0 < innerList.Count))
					{
						PropertyDescriptor pdValueMember;
						pdValueMember = this.m_currencyManager.GetItemProperties()[this.ValueMember];
						return pdValueMember.GetValue(innerList[index]);
					}
				}
				return null;
			}
			public object GetDisplay(int index)
			{
				IList innerList = this.m_currencyManager.List;
				if (innerList != null)
				{
					if ((this.DisplayMember != "") && (index >= 0 && 0 < innerList.Count))
					{
						PropertyDescriptor pdDisplayMember;
						pdDisplayMember = this.m_currencyManager.GetItemProperties()[this.ValueMember];
						return pdDisplayMember.GetValue(innerList[index]);
					}
				}
				return null;
			}
			
			#endregion
			
			#region Building the Radio Button List
			
			public void BuildRadioTable()
			{
				TableLayoutPanel radioTable = this.RadioTableLayout;
				radioTable.Controls.Clear();
				
				radioTable.RowStyles.Clear();
				radioTable.ColumnStyles.Clear();
				
				radioTable.RowStyles.Add(new RowStyle(SizeType.AutoSize));
				radioTable.ColumnStyles.Add(new ColumnStyle(SizeType.AutoSize));
				
				switch (this._radioLayoutStyle)
				{
					case RadioLayoutStyles.Horizontal:
						radioTable.ColumnCount = 1;
						radioTable.RowCount = this._RadioList.Count;
						break;
					case RadioLayoutStyles.Vertical:
						radioTable.ColumnCount = this._RadioList.Count;
						radioTable.RowCount = 1;
						break;
					case RadioLayoutStyles.Table:
						radioTable.ColumnCount = (this._RadioList.Count) / 2;
						radioTable.RowCount = radioTable.ColumnCount;
						break;
				}
				
				SPRadioButtonItem[,] tempItems = ArrangeItems(radioTable.RowCount, radioTable.ColumnCount);
				// Fill in the TableLayoutPanel.
				FillTable(tempItems, radioTable.RowCount, radioTable.ColumnCount, this.RadioTableLayout);
				
				
				
			}
			private SPRadioButtonItem[,] ArrangeItems(int rows, int cols)
			{
				
				// Return array of RadioButtonItem instances that matches
				// the layout of the control:
				SPRadioButtonItem[,] items = new SPRadioButtonItem[cols, rows];
				
				// Fill in the items array:
				int currentItem = 0;
				for (int col = 0; col <= cols - 1; col++)
				{
					for (int row = 0; row <= rows - 1; row++)
					{
						if (currentItem < this._RadioList.Count)
						{
							items[col, row] = this._RadioList[currentItem];
							currentItem++;
						}
					}
				}
				return items;
			}
			
			private void FillTable(SPRadioButtonItem[,] items, int rows, int cols, TableLayoutPanel tbl)
			{
				
				for (int col = 0; col <= cols - 1; col++)
				{
					for (int row = 0; row <= rows - 1; row++)
					{
						SPRadioButtonItem radioItem = items[col, row];
						if (radioItem != null)
						{
							if (! string.IsNullOrEmpty(radioItem.Name))
							{
								RadioButton btn = new RadioButton();
								btn.Name = radioItem.Name;
								btn.Text = radioItem.Text;
								btn.Dock = DockStyle.Fill;
								btn.CheckedChanged += new System.EventHandler(OnCheckedChanged);
								tbl.Controls.Add(btn, col, row);
							}
						}
					}
				}
			}
			
			#endregion
			
		}
		
		public class SPRadioButtonItemCollection : List<SPRadioButtonItem>
		{
			
			
			public SPRadioButtonItem this[string Name]
			{
				get
				{
					for (int cnt = 0; cnt <= this.Count - 1; cnt++)
					{
						if (this[cnt].Name == Name)
						{
							return this[cnt];
						}
					}
					
					return null;
				}
			}
			
		}
		public class SPRadioButtonItem
		{
			
			
			private string _Name = string.Empty;
			private bool _Checked = false;
			private int _ID = - 1;
			private string _Text = string.Empty;
			
			public string Name
			{
				get
				{
					return _Name;
				}
				set
				{
					_Name = value;
				}
			}
			public string Text
			{
				get
				{
					return _Text;
				}
				set
				{
					_Text = value;
				}
			}
			public bool Checked
			{
				get
				{
					return _Checked;
				}
				set
				{
					_Checked = value;
				}
			}
			
			internal SPRadioButtonItem()
			{
				
			}
			internal SPRadioButtonItem(string Name, string Text, bool Checked)
			{
				this.Name = Name;
				this.Text = Text;
				this.Checked = @Checked;
			}
		}
		
		#region Designer Extensions
		[System.Security.Permissions.PermissionSetAttribute(System.Security.Permissions.SecurityAction.Demand, Name = "FullTrust")]public class SPRadioButtonListDesigner : System.Windows.Forms.Design.ControlDesigner
		{
			
			
			private DesignerActionListCollection lists;
			
			//Use pull model to populate smart tag menu.
			public override DesignerActionListCollection ActionLists
			{
				get
				{
					if (lists == null)
					{
						lists = new DesignerActionListCollection();
						lists.Add(new SPRadioButtonListActionList(this.Component));
					}
					return lists;
				}
			}
		}
		
		/////////////////////////////////////////////////////////////////
		//DesignerActionList-derived class defines smart tag entries and
		// resultant actions.
		/////////////////////////////////////////////////////////////////
		public class SPRadioButtonListActionList : System.ComponentModel.Design.DesignerActionList
		{
			
			
			private SPRadioButtonList taskView;
			private DockStyle m_enmDockStyle = DockStyle.None;
			
			private DesignerActionUIService designerActionUISvc = null;
			
			//The constructor associates the control
			//with the smart tag list.
			public SPRadioButtonListActionList(IComponent component) : base(component)
			{
				
				this.taskView = (SPRadioButtonList) component;
				
				// Cache a reference to DesignerActionUIService, so the
				// DesigneractionList can be refreshed.
				this.designerActionUISvc = (DesignerActionUIService) (GetService(typeof(DesignerActionUIService)));
				
			}
			
			//Helper method to retrieve control properties. Use of
			// GetProperties enables undo and menu updates to work properly.
			private PropertyDescriptor GetPropertyByName(string propName)
			{
				PropertyDescriptor prop;
				prop = TypeDescriptor.GetProperties(taskView)[propName];
				if (prop == null)
				{
					throw (new ArgumentException("Matching SPRadioButtonList property not found!", propName));
				}
				else
				{
					return prop;
				}
			}
			
			public DockStyle Dock
			{
				get
				{
					return taskView.Dock;
				}
				set
				{
					GetPropertyByName("Dock").SetValue(taskView, value);
				}
			}
			
			
			//Method that is target of a DesignerActionMethodItem
			public void EditItems()
			{
				//GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
			}
			public void EditColumns()
			{
				//GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
			}
			
			public void EditGroups()
			{
				//GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
			}
			public void ParentDock()
			{
				if (Dock == DockStyle.Fill)
				{
					GetPropertyByName("Dock").SetValue(taskView, DockStyle.None);
				}
				else if (Dock == DockStyle.None)
				{
					GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill);
				}
			}
			
			//Implementation of this virtual method creates smart tag
			// items, associates their targets, and collects into list.
			public override DesignerActionItemCollection GetSortedActionItems()
			{
				DesignerActionItemCollection items = new DesignerActionItemCollection();
				
				//Define static section header entries.
				//items.Add(New DesignerActionHeaderItem("-"))
				items.Add(new DesignerActionHeaderItem(""));
				
				//items.Add(New DesignerActionMethodItem( _
				//Me, "EditItems", _
				//"Edit Items", _
				//" ", _
				//"Opens the Items Collection Editor"))
				
				//items.Add(New DesignerActionMethodItem( _
				//Me, "EditColumns", _
				//"Edit Columns", _
				//" ", _
				//"Opens the Columns Collection Editor"))
				
				//items.Add(New DesignerActionMethodItem( _
				//Me, "EditGroups", _
				//"Edit Groups", _
				//" ", _
				//"Opens the Groups Collection Editor"))
				
				if (Dock == DockStyle.None)
				{
					items.Add(new DesignerActionMethodItem(this, "ParentDock", "Dock in parent Container", "", "Dock in Parent Container"));
				}
				else if (Dock == DockStyle.Fill)
				{
					items.Add(new DesignerActionMethodItem(this, "ParentDock", "Undock from parent container", "", "Undock in Parent Container"));
					
				}
				return items;
			}
			
		}
		
		
		[System.Security.Permissions.PermissionSetAttribute(System.Security.Permissions.SecurityAction.Demand, Name = "FullTrust")]public class SPRadioButtonListItemsEditor : UITypeEditor
		{
			
			
			private IWindowsFormsEditorService editorService;
			
			public override object EditValue(ITypeDescriptorContext context, IServiceProvider provider, object value)
			{
				if ((context != null) && (context.Instance != null) && (provider != null))
				{
					editorService = (IWindowsFormsEditorService) (provider.GetService(typeof(IWindowsFormsEditorService)));
					
					if (editorService != null)
					{
						SPRadioButtonListEditorUI selectionControl = new SPRadioButtonListEditorUI(((SPRadioButtonItemCollection) value), editorService);
						
						editorService.ShowDialog(selectionControl);
						
						value = selectionControl.RadioItems;
					}
				}
				
				return value;
			}
			
			public override UITypeEditorEditStyle GetEditStyle(ITypeDescriptorContext context)
			{
				if ((context != null) && (context.Instance != null))
				{
					return UITypeEditorEditStyle.Modal;
				}
				return base.GetEditStyle(context);
			}
			
		}
		#endregion
	}
	
}

File: SPCode/UI/RadioButton/SPRadioButtonList.vb
Imports System.ComponentModel
Imports System.ComponentModel.Design
Imports System.Drawing.Design
Imports System.Windows.Forms
Imports System.Windows.Forms.Design
Imports System.Windows.Forms.ComponentModel

Namespace UI
    <Designer(GetType(SPRadioButtonListDesigner))> _
     <ToolboxBitmap(GetType(SPRadioButtonListDesigner))> _
     <DefaultEvent("SelectedIndexChanged")> _
    Public Class SPRadioButtonList

        Public Enum RadioLayoutStyles
            [Horizontal]
            [Vertical]
            [Table]
        End Enum

        Public Event SelectedIndexChanged(ByVal sender As Object, ByVal e As System.EventArgs)

        Private _RadioList As SPRadioButtonItemCollection = New SPRadioButtonItemCollection
        Private _radioLayoutStyle As RadioLayoutStyles
        Private _SelectedIndex As Integer = -1
        Private _SelectedValue As Object = Nothing

#Region "Overrides"
        Protected Overrides Sub OnLoad(ByVal e As System.EventArgs)
            MyBase.OnLoad(e)

            Try
                Me.RadioGroupBox.Text = Me.Text
                BuildRadioTable()
            Catch ex As Exception
            End Try
        End Sub

        Protected Sub OnCheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs)

        End Sub

#End Region

#Region "Properties"
        <Category("Behavior")> _
        <Description("Gets or Sets the Radio Buttons Layout Style")> _
        Public Property LayoutStyle() As RadioLayoutStyles
            Get
                Return _radioLayoutStyle
            End Get
            Set(ByVal value As RadioLayoutStyles)
                _radioLayoutStyle = value
            End Set
        End Property
        <Category("Behavior"), Browsable(True), _
        EditorAttribute(GetType(SPRadioButtonListItemsEditor), _
        GetType(System.Drawing.Design.UITypeEditor))> _
        Public Property Items() As SPRadioButtonItemCollection
            Get
                Return Me._RadioList
            End Get
            Set(ByVal value As SPRadioButtonItemCollection)
                Me._RadioList = value
                BuildRadioTable()
            End Set
        End Property
#End Region

#Region "Methods"
        Public Sub Add(ByVal Name As String, ByVal Text As String, ByVal Selected As Boolean)
            Me._RadioList.Add(New SPRadioButtonItem(Name, Text, Selected))
        End Sub

        Public Sub Remove(ByVal Name As String)
            Me._RadioList.Remove(Me._RadioList.Item(Name))
        End Sub
#End Region

#Region "Data Binding"
        Private m_currencyManager As CurrencyManager = Nothing
        Private m_ValueMember As String
        Private m_DisplayMember As String
        Private m_oDataSource As Object

        <Category("Data")> _
    Public Property DataSource() As Object
            Get
                Return m_oDataSource
            End Get
            Set(ByVal value As Object)
                If value Is Nothing Then
                    Me.m_currencyManager = Nothing
                    Me.Controls.Clear()
                Else
                    If Not (TypeOf value Is IList OrElse TypeOf m_oDataSource Is IListSource) Then
                        Throw (New System.Exception("Invalid DataSource"))
                    Else
                        If TypeOf value Is IListSource Then
                            Dim myListSource As IListSource = CType(value, IListSource)
                            If myListSource.ContainsListCollection = True Then
                                Throw (New System.Exception("Invalid DataSource"))
                            End If
                        End If
                        Me.m_oDataSource = value
                        Me.m_currencyManager = CType(Me.BindingContext(value), CurrencyManager)
                        BuildRadioTable()
                    End If
                End If
            End Set
        End Property ' end of DataSource property

        <Category("Data")> _
        Public Property ValueMember() As String
            Get
                Return Me.m_ValueMember
            End Get
            Set(ByVal value As String)
                Me.m_ValueMember = value
            End Set
        End Property
        <Category("Data")> _
        Public Property DisplayMember() As String
            Get
                Return Me.m_DisplayMember
            End Get
            Set(ByVal value As String)
                Me.m_DisplayMember = value
            End Set
        End Property
        Public Function GetValue(ByVal index As Integer) As Object
            Dim innerList As IList = Me.m_currencyManager.List
            If Not innerList Is Nothing Then
                If (Me.ValueMember <> "") AndAlso (index >= 0 AndAlso 0 < innerList.Count) Then
                    Dim pdValueMember As PropertyDescriptor
                    pdValueMember = Me.m_currencyManager.GetItemProperties()(Me.ValueMember)
                    Return pdValueMember.GetValue(innerList(index))
                End If
            End If
            Return Nothing
        End Function
        Public Function GetDisplay(ByVal index As Integer) As Object
            Dim innerList As IList = Me.m_currencyManager.List
            If Not innerList Is Nothing Then
                If (Me.DisplayMember <> "") AndAlso (index >= 0 AndAlso 0 < innerList.Count) Then
                    Dim pdDisplayMember As PropertyDescriptor
                    pdDisplayMember = Me.m_currencyManager.GetItemProperties()(Me.ValueMember)
                    Return pdDisplayMember.GetValue(innerList(index))
                End If
            End If
            Return Nothing
        End Function

#End Region

#Region "Building the Radio Button List"

        Public Sub BuildRadioTable()
            Dim radioTable As TableLayoutPanel = Me.RadioTableLayout
            radioTable.Controls.Clear()

            With radioTable
                .RowStyles.Clear()
                .ColumnStyles.Clear()

                .RowStyles.Add( _
                New RowStyle(SizeType.AutoSize))
                .ColumnStyles.Add( _
                 New ColumnStyle(SizeType.AutoSize))

                Select Case Me._radioLayoutStyle
                    Case RadioLayoutStyles.Horizontal
                        .ColumnCount = 1
                        .RowCount = Me._RadioList.Count
                    Case RadioLayoutStyles.Vertical
                        .ColumnCount = Me._RadioList.Count
                        .RowCount = 1
                    Case RadioLayoutStyles.Table
                        .ColumnCount = (Me._RadioList.Count) \ 2
                        .RowCount = .ColumnCount
                End Select

                Dim tempItems(,) As SPRadioButtonItem = ArrangeItems(.RowCount, .ColumnCount)
                ' Fill in the TableLayoutPanel.
                FillTable(tempItems, .RowCount, .ColumnCount, Me.RadioTableLayout)

            End With


        End Sub
        Private Function ArrangeItems( _
         ByVal rows As Integer, ByVal cols As Integer) _
         As SPRadioButtonItem(,)

            ' Return array of RadioButtonItem instances that matches
            ' the layout of the control:
            Dim items(cols - 1, rows - 1) As SPRadioButtonItem

            ' Fill in the items array:
            Dim currentItem As Integer = 0
            For col As Integer = 0 To cols - 1
                For row As Integer = 0 To rows - 1
                    If currentItem < Me._RadioList.Count Then
                        items(col, row) = Me._RadioList(currentItem)
                        currentItem += 1
                    End If
                Next
            Next
            Return items
        End Function

        Private Sub FillTable(ByVal items As SPRadioButtonItem(,), _
     ByVal rows As Integer, ByVal cols As Integer, _
     ByRef tbl As TableLayoutPanel)

            For col As Integer = 0 To cols - 1
                For row As Integer = 0 To rows - 1
                    Dim radioItem As SPRadioButtonItem = items(col, row)
                    If radioItem IsNot Nothing Then
                        If Not String.IsNullOrEmpty(radioItem.Name) Then
                            Dim btn As New RadioButton
                            btn.Name = radioItem.Name
                            btn.Text = radioItem.Text
                            btn.Dock = DockStyle.Fill
                            AddHandler btn.CheckedChanged, AddressOf OnCheckedChanged
                            tbl.Controls.Add(btn, col, row)
                        End If
                    End If
                Next
            Next
        End Sub

#End Region

    End Class

    Public Class SPRadioButtonItemCollection
        Inherits List(Of SPRadioButtonItem)

        Default Public Overloads ReadOnly Property Item(ByVal Name As String) As SPRadioButtonItem
            Get
                For cnt As Integer = 0 To Me.Count - 1
                    If Me.Item(cnt).Name = Name Then
                        Return Me.Item(cnt)
                    End If
                Next

                Return Nothing
            End Get
        End Property

    End Class
    Public Class SPRadioButtonItem

        Private _Name As String = String.Empty
        Private _Checked As Boolean = False
        Private _ID As Integer = -1
        Private _Text As String = String.Empty

        Public Property Name() As String
            Get
                Return _Name
            End Get
            Set(ByVal value As String)
                _Name = value
            End Set
        End Property
        Public Property Text() As String
            Get
                Return _Text
            End Get
            Set(ByVal value As String)
                _Text = value
            End Set
        End Property
        Public Property Checked() As Boolean
            Get
                Return _Checked
            End Get
            Set(ByVal value As Boolean)
                _Checked = value
            End Set
        End Property

        Friend Sub New()

        End Sub
        Friend Sub New(ByVal Name As String, ByVal Text As String, ByVal Checked As Boolean)
            Me.Name = Name
            Me.Text = Text
            Me.Checked = Checked
        End Sub
    End Class

#Region "Designer Extensions"
    <System.Security.Permissions.PermissionSetAttribute(System.Security.Permissions.SecurityAction.Demand, Name:="FullTrust")> _
        Public Class SPRadioButtonListDesigner
        Inherits System.Windows.Forms.Design.ControlDesigner

        Private lists As DesignerActionListCollection

        'Use pull model to populate smart tag menu.
        Public Overrides ReadOnly Property ActionLists() _
        As DesignerActionListCollection
            Get
                If lists Is Nothing Then
                    lists = New DesignerActionListCollection()
                    lists.Add( _
                    New SPRadioButtonListActionList(Me.Component))
                End If
                Return lists
            End Get
        End Property
    End Class

    '///////////////////////////////////////////////////////////////
    'DesignerActionList-derived class defines smart tag entries and
    ' resultant actions.
    '///////////////////////////////////////////////////////////////
    Public Class SPRadioButtonListActionList
        Inherits System.ComponentModel.Design.DesignerActionList

        Private taskView As SPRadioButtonList
        Private m_enmDockStyle As DockStyle = DockStyle.None

        Private designerActionUISvc As DesignerActionUIService = Nothing

        'The constructor associates the control 
        'with the smart tag list.
        Public Sub New(ByVal component As IComponent)

            MyBase.New(component)
            Me.taskView = CType(component, SPRadioButtonList)

            ' Cache a reference to DesignerActionUIService, so the
            ' DesigneractionList can be refreshed.
            Me.designerActionUISvc = _
            CType(GetService(GetType(DesignerActionUIService)),  _
            DesignerActionUIService)

        End Sub

        'Helper method to retrieve control properties. Use of 
        ' GetProperties enables undo and menu updates to work properly.
        Private Function GetPropertyByName(ByVal propName As String) _
        As PropertyDescriptor
            Dim prop As PropertyDescriptor
            prop = TypeDescriptor.GetProperties(taskView)(propName)
            If prop Is Nothing Then
                Throw New ArgumentException( _
                "Matching SPRadioButtonList property not found!", propName)
            Else
                Return prop
            End If
        End Function

        Public Property Dock() As DockStyle
            Get
                Return taskView.Dock
            End Get
            Set(ByVal value As DockStyle)
                GetPropertyByName("Dock").SetValue(taskView, value)
            End Set
        End Property


        'Method that is target of a DesignerActionMethodItem
        Public Sub EditItems()
            'GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
        End Sub
        Public Sub EditColumns()
            'GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
        End Sub

        Public Sub EditGroups()
            'GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
        End Sub
        Public Sub ParentDock()
            If Dock = DockStyle.Fill Then
                GetPropertyByName("Dock").SetValue(taskView, DockStyle.None)
            ElseIf Dock = DockStyle.None Then
                GetPropertyByName("Dock").SetValue(taskView, DockStyle.Fill)
            End If
        End Sub

        'Implementation of this virtual method creates smart tag  
        ' items, associates their targets, and collects into list.
        Public Overrides Function GetSortedActionItems() _
        As DesignerActionItemCollection
            Dim items As New DesignerActionItemCollection()

            'Define static section header entries.
            'items.Add(New DesignerActionHeaderItem("-"))
            items.Add(New DesignerActionHeaderItem(""))

            'items.Add(New DesignerActionMethodItem( _
            'Me, "EditItems", _
            '"Edit Items", _
            '" ", _
            '"Opens the Items Collection Editor"))

            'items.Add(New DesignerActionMethodItem( _
            'Me, "EditColumns", _
            '"Edit Columns", _
            '" ", _
            '"Opens the Columns Collection Editor"))

            'items.Add(New DesignerActionMethodItem( _
            'Me, "EditGroups", _
            '"Edit Groups", _
            '" ", _
            '"Opens the Groups Collection Editor"))

            If Dock = DockStyle.None Then
                items.Add(New DesignerActionMethodItem( _
                Me, "ParentDock", _
                 "Dock in parent Container", _
                "", _
                "Dock in Parent Container"))
            ElseIf Dock = DockStyle.Fill Then
                items.Add(New DesignerActionMethodItem( _
                Me, "ParentDock", _
                 "Undock from parent container", _
                "", _
                "Undock in Parent Container"))

            End If
            Return items
        End Function

    End Class


    <System.Security.Permissions.PermissionSetAttribute(System.Security.Permissions.SecurityAction.Demand, Name:="FullTrust")> _
    Public Class SPRadioButtonListItemsEditor
        Inherits UITypeEditor

        Private editorService As IWindowsFormsEditorService

        Public Overloads Overrides Function EditValue(ByVal context As ITypeDescriptorContext, ByVal provider As IServiceProvider, ByVal value As Object) As Object
            If ((context IsNot Nothing) And (context.Instance IsNot Nothing) And (provider IsNot Nothing)) Then
                editorService = CType(provider.GetService(GetType(IWindowsFormsEditorService)), IWindowsFormsEditorService)

                If (editorService IsNot Nothing) Then
                    Dim selectionControl As _
                          New SPRadioButtonListEditorUI( _
                          CType(value, SPRadioButtonItemCollection), _
                          editorService)

                    editorService.ShowDialog(selectionControl)

                    value = selectionControl.RadioItems
                End If
            End If

            Return value
        End Function

        Public Overloads Overrides Function GetEditStyle(ByVal context As ITypeDescriptorContext) As UITypeEditorEditStyle
            If ((context IsNot Nothing) And (context.Instance IsNot Nothing)) Then
                Return UITypeEditorEditStyle.Modal
            End If
            Return MyBase.GetEditStyle(context)
        End Function

    End Class
#End Region
End Namespace
